File: Trabalho 02/Código/main.py
def substituicoes_sucessivas(A, b):
    '''Executa o método das substituições sucessivas para resolver o sistema 
       linear triangular inferior Ax=b.
       Parâmetros de entrada: A é uma matriz triangular inferior e b é o vetor constante. 
       Saída: vetor x
    '''
    ## n é a ordem da matriz A
    n = len(A)
    
    x = n * [0]
    for i in range(0, n):
      S = 0
      for j in range(0, i):
        S = S + A[i][j] * x[j]
      x[i] = (b[i]-S)/A[i][i]
    
    return x

def identidade(n):
    '''Cria uma matriz identidade de ordem n.
    Parâmetros de entrada:  n é a ordem da matriz.
    Saída: matriz identidade de ordem n.
    '''
    m = []
    for i in range(0, n):
      linha = [0] * n
      linha[i] = 1
      m.append(linha)
    return m

def lu(A):
    '''
    Decompõe a matriz A no produto de duas matrizes L e U. Onde L é uma matriz 
    triangular inferior unitária e U é uma matriz triangular superior.
    Parâmetros de entrada: A é uma matriz quadrada de ordem n.
    Saída: (L,U) tupla com as matrizes L e U
    '''
    n = len(A)
    
    ## Inicializa a matriz L com a matriz identidade
    L = identidade(n)
    
    for k in range(0, n-1):
      for i in range(k+1, n):
        m = - A[i][k]/A[k][k]
        L[i][k] = -m

        for j in range(k+1, n):
          A[i][j] = m * A[k][j] + A[i][j]

          A[i][k] = 0
    
    return (L, A)

def formata_matriz(M):
    m = len(M) # número de linhas
    n = len(M[0]) # número de colunas
    s = ""
    for i in range(m):
        for j in range(n):
           s += "%9.3f " % M[i][j]
        s += "\n"
    return s

def substituicoes_retroativas(A, b):
    '''Executa o método das substituições retroativas para resolver o sistema 
       linear triangular superior Ax=b.
       Parâmetros de entrada: A é uma matriz triangular superior e b é o vetor constante. 
    '''
    ## n é a ordem da matriz A
    n = len(A)
    
    ## inicializa o vetor x com tamanho n e elementos iguais a 0
    x = n * [0] 
    
    for i in range(n-1, -1, -1):
      S = 0
      for j in range(i+1, n):
        S = S + A[i][j] * x[j]
      x[i] = (b[i]-S)/A[i][i]
    
    return x

def lux(L,U,b):
    '''
    Resolve o sistema LUx=b.
    Esse método resolve os dois sistemas lineares triangulares.
    Parâmetros de entrada: L é uma matriz triangular inferior de ordem n,
    U é uma matriz triangular superior e b é o vetor constante.
    Saída: vetor x solução do sistema.
    '''
    
    y = substituicoes_sucessivas(L, b)

    x = substituicoes_retroativas(U, y)
    
    return x

def ldp(A):
    '''
    Decompõe a matriz A no produto de três matrizes L, D e P. 
    Onde L é uma matriz triangular inferior unitária, 
    D é uma matriz diagonal e P é uma matriz triangular superior com diagonal unitária (U = DP).
    
    Parâmetros de entrada: A é uma matriz quadrada de ordem n.
    Saída: (L, D, P) tupla com as matrizes L, D e P.
    '''
    n = len(A)
    
    L = identidade(n)
    D = identidade(n)
    
    for k in range(0, n-1):
        for i in range(k+1, n):
            m = -A[i][k] / A[k][k]
            
            # Atualiza a matriz L
            L[i][k] = -m
            
            for j in range(k+1, n):
                A[i][j] = m * A[k][j] + A[i][j]
            
                A[i][k] = 0
    
    # Atualiza a matriz D
    for i in range(n):
        D[i][i] = A[i][i]
        
        # Atualiza a matriz P
        pivot = A[i][i]
        for j in range(i, n):
            A[i][j] = A[i][j] / pivot

    
    return (L, D, A)

def ldp_solver(L, D, P, b):
    '''
    Resolve o sistema LDPx=b.
    Esse método resolve os três sistemas lineares.
    Parâmetros de entrada: L é uma matriz triangular inferior de ordem n,
    D é uma matriz diagonal, P é uma matriz triangular superior com diagonal unitária 
    e b é o vetor constante.
    Saída: vetor x solução do sistema.
    '''
    # Resolve Ls = b
    s = substituicoes_sucessivas(L, b)
    
    # Resolve Dy = s
    y = [s[i] / D[i][i] for i in range(len(D))]
    
    # Resolve Px = y
    x = substituicoes_retroativas(P, y)
    
    return x

def main():
    A1 = [[3, -2, 1],
    [1, -3, 4],
    [9, 4, -5]]

    b = [8, 6, 11]

    (L, U) = lu(A1)
    x1 = lux(L, U, b)
    print("Matrix L:")
    print(formata_matriz(L))
    print("Matrix U:")
    print(formata_matriz(U))
    print(x1)
    
    
    A2 = [[3, -2, 1],
    [1, -3, 4],
    [9, 4, -5]]
    

    (L, D, P) = ldp(A2)
    x2 = ldp_solver(L, D, P, b)
    
    print("Matrix L:")
    print(formata_matriz(L))
    print("Matrix D:")
    print(formata_matriz(D))
    print("Matrix P:")
    print(formata_matriz(P))
    print("Solution x:")
    print(x2)

if __name__ == "__main__":
    main()

File: Trabalho 02/Código/main.c
#include <stdio.h>
#include <stdlib.h>

void substituicoes_sucessivas(double **A, double *b, double *x, int n) {
    for (int i = 0; i < n; i++) {
        double S = 0;
        for (int j = 0; j < i; j++) {
            S += A[i][j] * x[j];
        }
        x[i] = (b[i] - S) / A[i][i];
    }
}

void substituicoes_retroativas(double **A, double *b, double *x, int n) {
    for (int i = n - 1; i >= 0; i--) {
        double S = 0;
        for (int j = i + 1; j < n; j++) {
            S += A[i][j] * x[j];
        }
        x[i] = (b[i] - S) / A[i][i];
    }
}

double** identidade(int n) {
    double **m = (double **)malloc(n * sizeof(double *));
    for (int i = 0; i < n; i++) {
        m[i] = (double *)malloc(n * sizeof(double));
        for (int j = 0; j < n; j++) {
            m[i][j] = (i == j) ? 1.0 : 0.0;
        }
    }
    return m;
}

void lu(double **A, double ***L, double ***U, int n) {
    *L = identidade(n);

    for (int k = 0; k < n - 1; k++) {
        for (int i = k + 1; i < n; i++) {
            double m = -A[i][k] / A[k][k];
            (*L)[i][k] = -m;

            for (int j = k + 1; j < n; j++) {
                A[i][j] = m * A[k][j] + A[i][j];
            }

            A[i][k] = 0;
        }
    }

    *U = A;
}

void lux(double **L, double **U, double *b, double *x, int n) {
    double *y = (double *)malloc(n * sizeof(double));

    substituicoes_sucessivas(L, b, y, n);
    substituicoes_retroativas(U, y, x, n);

    free(y);
}

void ldp(double **A, double ***L, double ***D, double ***P, int n) {
    *L = identidade(n);
    *D = (double **)malloc(n * sizeof(double *));

    for (int k = 0; k < n - 1; k++) {
        for (int i = k + 1; i < n; i++) {
            double m = -A[i][k] / A[k][k];

            // Atualiza a matriz L
            (*L)[i][k] = -m;

            for (int j = k + 1; j < n; j++) {
                A[i][j] = m * A[k][j] + A[i][j];
            }

            A[i][k] = 0;
        }
    }

  for (int i = 0; i < n; i++) {
      (*D)[i] = (double *)malloc(n * sizeof(double));

      for (int j = 0; j < n; j++) {
          (*D)[i][j] = (i == j) ? A[i][i] : 0.0;
      }

      double pivot = A[i][i];
      for (int j = i; j < n; j++) {
          A[i][j] = A[i][j] / pivot;
      }
  }

    *P = A;
}

void ldp_solver(double **L, double **D, double **P, double *b, double *x, int n) {
    double *s = (double *)malloc(n * sizeof(double));
    double *y = (double *)malloc(n * sizeof(double));

    substituicoes_sucessivas(L, b, s, n);

    for (int i = 0; i < n; i++) {
        y[i] = s[i] / D[i][i];
    }

    substituicoes_retroativas(P, y, x, n);

    free(s);
    free(y);
}

void imprimir_matriz(double **matriz, int m, int n) {
    for (int i = 0; i < m; i++) {
        for (int j = 0; j < n; j++) {
            printf("%9.3f ", matriz[i][j]);
        }
        printf("\n");
    }
}

int main() {
    
  int n;
  printf("Informe a ordem da matriz: ");
  scanf("%d", &n);

  // Alocar dinamicamente a matriz A
  double **A1 = (double **)malloc(n * sizeof(double *));
  for (int i = 0; i < n; i++) {
      A1[i] = (double *)malloc(n * sizeof(double));
  }

  // Alocar dinamicamente o vetor f
  double *f1 = (double *)malloc(n * sizeof(double));

  // Preencher a matriz A
  printf("Informe os elementos da matriz A:\n");
  for (int i = 0; i < n; i++) {
      for (int j = 0; j < n; j++) {
          printf("A[%d][%d]: ", i + 1, j + 1);
          scanf("%lf", &A1[i][j]);
      }
  }

  // Preencher o vetor b
  printf("\nInforme os elementos do vetor f:\n");
  for (int i = 0; i < n; i++) {
      printf("f[%d]: ", i + 1);
      scanf("%lf", &f1[i]);
  }
  
  double** A2 = (double**)malloc(n * sizeof(double*));
  for (int i = 0; i < n; i++) {
        A2[i] = (double*)malloc(n * sizeof(double));
      for (int j = 0; j < n; j++) {
          A2[i][j] = A1[i][j];
      }
  }

  double* f2 = (double*)malloc(n * sizeof(double));
  for (int i = 0; i < n; i++) {
      f2[i] = f1[i];
  }

  printf("\nMétodo de Fatoração LU normal:\n");
  double **L1, **U1, **D1;
  double *d1 = (double *)malloc(n * sizeof(double));
  lu(A1, &L1, &U1, n);
  /*
  printf("Matrix L:\n");
  imprimir_matriz(L1, n, n);
  printf("Matrix U:\n");
  imprimir_matriz(U1, n, n);
  */
  lux(L1, U1, f1, d1, n);
  printf("Vetor d:\n");
  for (int i = 0; i < n; i++) {
      printf("%9.3f\n", d1[i]);
  }

  printf("\nMétodo de Fatoração LDP:\n");
  double **L2, **U2, **D2, **P2;
  double *d2 = (double *)malloc(n * sizeof(double));
  ldp(A2, &L2, &D2, &P2, n);
  /*
  printf("Matrix L:\n");
  imprimir_matriz(L2, n, n);
  printf("Matrix D:\n");
  imprimir_matriz(D2, n, n);
  printf("Matrix P:\n");
  imprimir_matriz(P2, n, n);
  */
  
  ldp_solver(L2, D2, P2, f2, d2, n);
  printf("Vetor d:\n");
  for (int i = 0; i < n; i++) {
      printf("%9.3f\n", d2[i]);
  }

  return 0;
}
